src/components/Taxonomy/Taxonomy.tsx

```diff
@@ -454,14 +451,47 @@ const Taxonomy = ({
   }, [externalSelected]);
 
   useEffect(() => {
-    if (isOpen) {
-      document.body.addEventListener("click", onClickOutside, true);
-      document.body.addEventListener("keydown", onEsc);
-    } else {
+    document.body.addEventListener("click", onClickOutside, true);
+    document.body.addEventListener("keydown", onKeyDown);
+    return () => {
       document.body.removeEventListener("click", onClickOutside);
-      document.body.removeEventListener("keydown", onEsc);
-    }
-  }, [isOpen]);
+      document.body.removeEventListener("keydown", onKeyDown);
+    };
+  }, []);
+
+  const onKeyDown = useCallback(e => {
+    const taxonomyList: NodeListOf<HTMLElement> | undefined = taxonomyRef.current?.querySelectorAll('.item');
+    const searchInput = taxonomyRef.current?.querySelector("input");
+    const focusedElement: HTMLInputElement | Element | any  = document.activeElement || undefined;
+    const taxonomyHasItems = taxonomyList && taxonomyList.length > 0;
+    const index = (taxonomyList && focusedElement) ? Array.from(taxonomyList).findIndex((taxonomyItem => taxonomyItem.id === focusedElement.id)) : -1;
+    const shiftFocus = (index: number, shift: number) => taxonomyHasItems && taxonomyList[index + shift].focus();
+
+    switch (e.key) {
+      case "Escape":
+        close();
+        e.stopPropagation();
+        break;
+      case "ArrowDown":
+        if (e.shiftKey) {
+          setOpen(true);
+          searchInput && searchInput.focus();
+        }
+        if (index >= 0) shiftFocus(index, 1);
+        if (searchInput === focusedElement) shiftFocus(0, 0);
+        break;
+      case "ArrowUp":
+        if (index > 0) shiftFocus(index, -1);
+        else if (index === 0) searchInput && searchInput.focus();
+        break;
+      case "Shift":
+        if (index >= 0) focusedElement.parentNode?.parentNode?.toggle(focusedElement.id);
+        if (searchInput) searchInput.focus();
```

Comment:
why do we focus right after opening the subtree?

Reply:
the focus is lost when the sub tree is re rendered. I'm open to suggestions